lib/collection/src/lib.rs

```diff
@@ -411,20 +432,40 @@ impl Collection {
         let limit = limit + 1;
 
         let mut points = Vec::new();
-        for shard in self.all_shards() {
-            let mut shard_points = shard
-                .get()
-                .scroll_by(
-                    segment_searcher,
-                    offset,
-                    limit,
-                    &with_payload_interface,
-                    with_vector,
-                    request.filter.as_ref(),
-                )
-                .await?;
-            points.append(&mut shard_points);
-        }
+
+        match shard_selection {
+            None => {
+                for shard in self.all_shards() {
+                    let mut shard_points = shard
+                        .get()
+                        .scroll_by(
+                            segment_searcher,
+                            offset,
+                            limit,
+                            &with_payload_interface,
+                            with_vector,
+                            request.filter.as_ref(),
+                        )
+                        .await?;
+                    points.append(&mut shard_points);
+                }
+            }
+            Some(shard_selection) => {
+                let local_shard = self.local_shard_by_id(shard_selection)?;
+                let mut shard_points = local_shard
+                    .scroll_by(
+                        segment_searcher,
+                        offset,
+                        limit,
+                        &with_payload_interface,
+                        with_vector,
+                        request.filter.as_ref(),
+                    )
+                    .await?;
+                points.append(&mut shard_points);
+            }
+        };
```

Comment:
It looks like it should be possible to reduce code duplication by selecting shards first. And the making request on whatever shards selected.

Reply:
I tried to create an Iterator with a single element in case of shard selection to have minimum code duplication but I struggled to make it compile.
I will give it another go.

Reply:
done in https://github.com/qdrant/qdrant/pull/463/commits/8a5dfcdb6fbda5e4dec49f1d28dd9fa62c27c73a